src/meili/match_route.cc

```diff
@@ -129,38 +134,47 @@ bool MergeRoute(const State& source,
 void cut_segments(const std::vector<MatchResult>& match_results,
                   int first_idx,
                   int last_idx,
-                  const std::vector<EdgeSegment>& segments,
+                  std::vector<EdgeSegment>& segments,
                   std::vector<EdgeSegment>& new_segments) {
   auto first_segment = segments.begin();
+  auto search_segment = first_segment;
   int prev_idx = first_idx;
   for (int curr_idx = first_idx + 1; curr_idx <= last_idx; ++curr_idx) {
     const MatchResult& curr_match = match_results[curr_idx];
     const MatchResult& prev_match = match_results[prev_idx];
 
-    // skip if we do not need to cut
-    if (!curr_match.is_break_point && curr_idx != last_idx) {
-      continue;
-    }
-
-    // Allow for some fp-fuzz in this gt comparison
+    // does the previous match point look further along its edge than the current
     bool prev_gt_curr = prev_match.distance_along > curr_match.distance_along + 1e-3;
 
-    // we want to handle to loop by locating the correct target edge by comparing the distance alone
+    // if they are on the same edge and previous comes after then its a loop
     bool loop = (prev_match.edgeid == curr_match.edgeid) && prev_gt_curr;
 
     // if it is a loop, we start the search after the first edge
-    auto last_segment = std::find_if(first_segment + static_cast<size_t>(loop), segments.end(),
-                                     [&curr_match](const EdgeSegment& segment) {
-                                       return (segment.edgeid == curr_match.edgeid);
-                                     });
+    for (search_segment += static_cast<size_t>(loop); search_segment != segments.end();
+         ++search_segment) {
+      if (search_segment->edgeid == curr_match.edgeid)
+        break;
+    }
 
-    if (last_segment == segments.cend()) {
+    if (search_segment == segments.cend()) {
       throw std::logic_error("In meili::cutsegments(), unexpectedly unable to locate target edge.");
     }
 
+    // if its not a break point nor the last segment, then we dont need to split it into two
+    if (!curr_match.is_break_point && curr_idx != last_idx) {
+      // but if it isnt a breakpoint and its not the last stateful match point then its interpolated
+      // and we still need to mark the segments in such a way that we which match points lie on them
+      if (search_segment->first_match_idx < 0)
+        search_segment->first_match_idx = curr_idx;
+      if (search_segment->last_match_idx < 0)
+        search_segment->last_match_idx = curr_idx;
```

Comment:
i'll admit that i didnt think too hard about how correct this is but rather did the conservative thing, that is only set the index if unset. its possible that i should just always set both of them because they will be corrected below and wont have been set up to this point (ie we can probably remove the `if`s) but i didnt want to have to prove that to myself at 10pm :smile: